File: leetcode/125-valid-palindrome.js
// https://leetcode.com/problems/valid-palindrome/ (top-interview-questions)

/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    
  let l = 0;
  let r = s.length - 1;

  while (l <= r) {
    
    if (!/^[a-zA-Z0-9]*$/.test(s.charAt(l))) {
      l += 1;
      continue;
    };

    if (!/^[a-zA-Z0-9]*$/.test(s.charAt(r))) {
      r -= 1;
      continue;
    };

    if (s.charAt(l).toLowerCase() !== s.charAt(r).toLowerCase()) return false;
    
    l += 1;
    r -= 1;
  }

  return true;
  
};

File: leetcode/125-valid-palindrome.py
# https://leetcode.com/problems/valid-palindrome/ (top-interview-questions)

class Solution:
    def isPalindrome(self, s: str) -> bool:
        left_ptr, right_ptr = 0, len(s)-1
        
        while left_ptr < right_ptr:
            while left_ptr < right_ptr and not s[left_ptr].isalnum():
                left_ptr += 1
            while right_ptr > left_ptr and not s[right_ptr].isalnum():
                right_ptr -= 1

            if  s[left_ptr].lower() != s[right_ptr].lower():
                return False
            
            left_ptr += 1
            right_ptr -= 1
            
        return True
